Fixes An Issue With Every Hour Cron Schedules

NeilBetham opened the pull request:
I noticed an issue where if I have a schedule set to run every hour it will get overridden in favor of a schedule that only gets run at a specific hour that is before the current time.

The logic in schedule.py uses the start of the day as the starting point for the croniter. As such events that run hourly appear to happen before events that run at a specific hour of the day. To fix this I added a helper which indexes the croniter backwards from the current time instead of forwards from the start of the day. In my local testing this seems to have resolved the issue.

maxbbraun commented:
Thanks for looking into this, Neil! Could you give an example of cron expressions together with the expected and actual behavior for those?

NeilBetham commented:
Sure! So for example say we have two crons in the schedule:

1. `20 * * * *` 
2. `10 17 * * *`

In this setup the way the schedule is evaluated currently is by setting the croniter start time to 00:00 and then calling next. This will yield the following times for each schedule respectively:

1. `00:20`
2. `17:10`

Now if the current time is before `17:10` then the first schedule is the only time to make it through the filter for events in the past. However say for example the time is currently `17:25` and the client is calling the API for a new image then both schedules pass that filter. Then once they are run through the max filter [here](https://github.com/maxbbraun/accent/blob/master/server/schedule.py#L125) then the second schedule is selected because the first schedule is only evaluated at `00:20` which puts it before `17:10` in the max filter.

My thinking is that the correct behavior would be to step back in time from the current instant and see when each of the schedules would have most recently executed in the past and then run that same max filters on those instances so that the most recently passed event will be selected. With this PR the times that would evaluate from the example schedule would be the following:

1. `17:20`
2. `17:10`

After the max function the first schedule would be correctly selected.

Let me know if anything is still not clear!

maxbbraun commented on server/schedule.py:
```diff
@@ -85,6 +85,19 @@ def _next(self, cron, after, user):
             raise ContentError(e)
 
+    def _prev(seld, cron, before, user):
+        """Finds the previous time matching the cron expression"""
```
```suggestion
        """Finds the previous time matching the cron expression."""
```

NeilBetham replied:
Fixed!

maxbbraun commented on server/schedule.py:
```diff
@@ -85,6 +85,19 @@ def _next(self, cron, after, user):
         except ValueError as e:
             raise ContentError(e)
 
+    def _prev(seld, cron, before, user):
```
```suggestion
    def _previous(self, cron, before, user):
```

NeilBetham replied:
Fixed!

maxbbraun commented on server/schedule.py:
```diff
@@ -85,6 +85,19 @@ def _next(self, cron, after, user):
+        """Finds the previous time matching the cron expression"""
+
+        try:
+            cron = seld._sun.rewrite_cron(cron, before, user)
```
```suggestion
            cron = self._sun.rewrite_cron(cron, before, user)
```

NeilBetham replied:
Well this one's embarrassing. Fixed.

maxbbraun commented on server/schedule.py:
```diff
@@ -85,6 +85,19 @@ def _next(self, cron, after, user):
+        try:
+            cron = seld._sun.rewrite_cron(cron, before, user)
+        except DataError as e:
+            raise ContentError
```
```suggestion
            raise ContentError(e)
```

NeilBetham replied:
Fixed!

maxbbraun commented on server/schedule.py:
```diff
@@ -114,7 +127,7 @@ def image(self, user, width, height):
         today = time.replace(hour=0, minute=0, second=0, microsecond=0)
         while True:
-            entries = [(self._next(entry['start'], today, user), entry)
+            entries = [(self._prev(entry['start'], time, user), entry)
```
```suggestion
            entries = [(self._previous(entry['start'], time, user), entry)
```

NeilBetham replied:
Fixed!

maxbbraun commented on server/schedule.py:
```diff
@@ -114,7 +127,7 @@ def image(self, user, width, height):
             raise ContentError(e)
         today = time.replace(hour=0, minute=0, second=0, microsecond=0)
```
```suggestion
```

NeilBetham replied:
Fixed!

maxbbraun reviewed (changes requested):
Thanks! Left a few suggestions.

I would also suggest including [`sun.py`](https://github.com/maxbbraun/accent/blob/master/server/sun.py) in the PR, because the [`rewrite_cron`](https://github.com/maxbbraun/accent/blob/60767095a051cb6984b91ab670c356597a56623f/server/sun.py#L20) method there is now being used differently. At a minimum, the `after` variable should be renamed to something like `reference`, because it is now used in both directions. We should also verify that the sunset/sunrise logic there still works as intended when going backwards in time.

NeilBetham commented:
Thanks for the review. I've address the comments and suggestions you made. I have updated sun.py to more align with how it worked previously. Basically I moved the munging of the reference time to midnight from the `image()` method to the `rewrite_cron` method. This seems to replicate the original behavior; though I'm not 100% clear on what the expected behavior for `rewrite_cron` is so please correct me if I'm wrong here. I have tested this locally the following logs are with the home address set to `San Francisco, CA`

#### `/gif`
```
INFO:root:Rewrote cron: (sunrise * * *) -> (10 6 * * *), reference Tuesday May 04 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (2 20 * * *), reference Tuesday May 04 2021 00:00:00 PDT 
INFO:root:Using image from schedule entry: City Sunset (sunset * * *, Tuesday May 04 2021 20:02:00 PDT)
```

#### Schedule Render
```
INFO:root:Rewrote cron: (sunrise * * *) -> (11 6 * * *), reference Monday May 03 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (1 20 * * *), reference Monday May 03 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (11 6 * * *), reference Monday May 03 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (1 20 * * *), reference Monday May 03 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (11 6 * * *), reference Monday May 03 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (1 20 * * *), reference Monday May 03 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (10 6 * * *), reference Tuesday May 04 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (2 20 * * *), reference Tuesday May 04 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (10 6 * * *), reference Tuesday May 04 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (2 20 * * *), reference Tuesday May 04 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (9 6 * * *), reference Wednesday May 05 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (3 20 * * *), reference Wednesday May 05 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (9 6 * * *), reference Wednesday May 05 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (3 20 * * *), reference Wednesday May 05 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (8 6 * * *), reference Thursday May 06 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (4 20 * * *), reference Thursday May 06 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (8 6 * * *), reference Thursday May 06 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (4 20 * * *), reference Thursday May 06 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (7 6 * * *), reference Friday May 07 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (5 20 * * *), reference Friday May 07 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (7 6 * * *), reference Friday May 07 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (5 20 * * *), reference Friday May 07 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (6 6 * * *), reference Saturday May 08 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (6 20 * * *), reference Saturday May 08 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (6 6 * * *), reference Saturday May 08 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (6 20 * * *), reference Saturday May 08 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (5 6 * * *), reference Sunday May 09 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (7 20 * * *), reference Sunday May 09 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunrise * * *) -> (5 6 * * *), reference Sunday May 09 2021 00:00:00 PDT 
INFO:root:Rewrote cron: (sunset * * *) -> (7 20 * * *), reference Sunday May 09 2021 00:00:00 PDT 
```

maxbbraun commented on server/sun.py:
```diff
@@ -17,16 +17,19 @@ def __init__(self, geocoder):
             return cron
 
+        # Replace H:M:S in the reference time so that we get the correct sunrise / set
+        reference_midnight = reference.replace(hour=0, minute=0, second=0, microsecond=0)
```
Not sure I understand this one yet. What happens if you don't do this?

NeilBetham replied:
So the logic to pick the most recent sunrise / sunset seems to depend on the input datetime being midnight for the current day. I checked the output times without the `replace` and they weren't consistent with the implementation that exists in master.

maxbbraun replied:
Hm, it always takes me a while to wrap my head around this logic...

If I understand this diff correctly, it modifies the meaning of `next_sunrise` and `next_sunset` below to find the *first* sunrise and sunset within the reference day rather than the *next* one after the reference time (which was the previous behavior).

Unless I'm misunderstanding something, this does seem to introduce some issues. For example, when `schedule.delay()` requests the next time to wake up sometime during the day and the entry references `sunset`, then it may get a time in the past.

NeilBetham replied:
Ahh yeah, you are correct. I missed the interdependence with `delay()`. I will fix that.

maxbbraun reviewed (commented):
Also, please run [`pycodestyle`](https://pypi.org/project/pycodestyle/) and fix any issues that come up.

knyar commented on server/schedule.py:
```diff
@@ -85,6 +85,19 @@ def _next(self, cron, after, user):
+            raise ContentError(e)
+
+        try:
+            return croniter(cron, before, user).get_prev(datetime)
```
Passing `user` as the third argument to `croniter` looks accidental here.

NeilBetham replied:
Ahh yeah, good catch. Not sure what I was thinking.

knyar commented on server/schedule.py:
```diff
@@ -112,9 +125,9 @@ def image(self, user, width, height):
             raise ContentError(e)
-        today = time.replace(hour=0, minute=0, second=0, microsecond=0)
+
         while True:
```
I don't think you need this loop anymore: `get_prev` should look back more than 1 day, so I don't think you need to manually keep decrease `time`.

NeilBetham replied:
Yup all that makes sense. Thanks!

knyar reviewed (commented):
Perhaps it's also worth updating `croniter` to a more modern version?

NeilBetham commented:
@knyar Is there something fixed in `croniter` that is broken in the current version that we use?

NeilBetham commented:
Okay, updated based on `pydocstyle` feedback as well as feedback here. Please let me know if there is anything else you'd like changed!

knyar commented:
> @knyar Is there something fixed in croniter that is broken in the current version that is used?

I am not aware of anything specific, I just noticed that the version listed in `requirements.txt` is pretty old.

knyar commented:
Just as a data point, I've been running this on my personal accent server for a couple of weeks, and it seems to work fine. Unlike the current version it correctly handles my very simple schedule: cycling between two images every hour.

Hopefully this feedback might help to make a decision to get it merged.

maxbbraun reviewed (commented):
Sorry I haven't been able to reply more quickly!

I left a few more comments and it looks like there are also still some `pycodestyle` issues:

```
schedule.py:72:80: E501 line too long (83 > 79 characters)
sun.py:20:80: E501 line too long (81 > 79 characters)
sun.py:31:80: E501 line too long (86 > 79 characters)
sun.py:32:80: E501 line too long (89 > 79 characters)
sun.py:55:80: E501 line too long (83 > 79 characters)
sun.py:69:80: E501 line too long (81 > 79 characters)
```

NeilBetham commented on server/main.py:
```diff
@@ -269,4 +269,7 @@ def server_error(e):
 
 
 if __name__ == '__main__':
+    import logging
```
Oh yeah, I was using it for local testing. Will remove.

NeilBetham replied:
Fixed!

NeilBetham commented on server/schedule.py:
```diff
@@ -73,8 +80,7 @@ def __init__(self, geocoder):
     def _next(self, cron, after, user):
-        """Finds the next time matching the cron expression."""
-
+        """Find the next time matching the cron expression."""
```
So I was using the most recent release of pycodestyle and these were the corrections it was asking me to make. Do you have a config that could be added to the repo to match the settings you are using?

NeilBetham replied:
Reverted all the formatting changes and have just addressed the ones from `pycodestyle`

NeilBetham commented on server/schedule.py:
```diff
@@ -50,7 +50,9 @@
 
 class Schedule(ImageContent):
-    """A database-backed schedule determining which images to show at request
+    """User schedule processing.
```
Yeah, in this case pydocstyle was complaining about the formatting and that it didn't have a description.

NeilBetham commented on server/schedule.py:
```diff
@@ -85,9 +91,20 @@ def _next(self, cron, after, user):
-    def _image(self, kind, user, width, height):
-        """Creates an image based on the kind."""
+    def _previous(self, cron, before, user):
+        """Find the previous time matching the cron expression."""
```
Same as above, I'm good with any syntax for these. I was just trying to appease pydocstyle here as well.

NeilBetham commented on server/schedule.py:
```diff
@@ -105,29 +122,21 @@ def _image(self, kind, user, width, height):
+
+        # Use the most recent past entry.
+        latest_datetime, latest_entry = max(entries,
+                                            key=lambda x: x[0])
```
Yeah for sure.

NeilBetham replied:
Fixed!

NeilBetham commented:
Yeah, no worries on the time front! These are side projects so I understand IRL takes precedence. Most of the changes to docs here were just to make `pydocstyle` happy. I don't have a preference one way or another on them. I can revert the changes I made and then make fixes based on your config if you prefer.

maxbbraun commented:
Yeah, let's keep it to the functional changes for now. The `pycodestyle` I ran doesn't seem to do docstring and just helps with stuff [like this](https://github.com/maxbbraun/accent/pull/19#pullrequestreview-669667489).

NeilBetham commented:
Weird, I can't repro the extra critiques it was giving me anymore. Oh well. I'll revert those and fix the issue with `sun.py` / `delay()`. Thanks for the feedback!

NeilBetham commented:
Alright, sorry for the delay. This time around I have reverted all the doc changes and have updated `sun.py` to support the same forward and backward semantics as `schedule.py`. This should make sure that sunrise / sunset rewriting works as expected. I have done some testing locally and everything seems to be working as expected.
